### PR TITLE
OIDC: stop modifying deprecated fields and setters on former config classes marked for removal

### PR DESCRIPTION
**Pull request by michalvavrik**
In Quarkus OIDC we use deprecated setters and modify fields on former configuration classses. These methods and fields are deprecated and marked for removal. We don't want users to use them. Ideally we want to consume SmallRye Config output and only allow modification of properties that depend on dynamic information (is OIDC provider available? does it support user information?). This PR:

- limits modification of "tenant config" to 2 changes (user info and tenant enabled, both are called once at max)
- avoid using setters and fields that are deprecated and marked for removal
- gives Quarkus OIDC control over tenant config state (AKA config is immutable except for the 2 places mentioned above):
  - static tenant config is final
  - dynamic tenant config can only by changed from `TenantConfigResolver`, not anywhere else and users cannot change config state used in a different HTTP requests (which happened in some of our tests)

Later, we remove deprecated methods and fields, and only Quarkus OIDC will be able to modify small number of fields and stop instantiating great lot of former configuration classes.

### CODE REVIEW

**Comment by michalvavrik** [comment]
2 of the failures seems related, I'll have a look.

**Comment by michalvavrik** [comment]
Ah, I think I got it, uncovered bit of a flaw in tenant initialization, I just hope there wasn't a good reason why it was there. Let's see how tests go.

**Comment by michalvavrik** [comment]
The fact that config is created on multiple places even before it is modified makes things much harder. I think we need to keep at least properties that are set based on resolved metadata mutable (which is 1 property AFAIK) and the rest must be set as soon as config is built. The config class instance must be one, not rebuilt.

**Comment by michalvavrik** [comment]
I'll play with it next week.

File: extensions/oidc/runtime/src/test/java/io/quarkus/oidc/runtime/KnownOidcProvidersTest.java [hunk 1]
```diff
@@ -513,7 +533,8 @@ public void testOverrideTwitchProperties() throws Exception {
         tenant.credentials.clientSecret.setMethod(Method.BASIC);
         tenant.authentication.setForceRedirectHttpsScheme(false);
 
-        OidcTenantConfig config = OidcUtils.mergeTenantConfig(tenant, KnownOidcProviders.provider(Provider.FACEBOOK));
+        tenant.setProvider(Provider.TWITCH);
```
Comment (michalvavrik) [inline extensions/oidc/runtime/src/test/java/io/quarkus/oidc/runtime/KnownOidcProvidersTest.java:536]: This is intentional switch from `FACEBOOK` to `TWITCH`, I think the previous state was a typo.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java [hunk 2]
```diff
@@ -1239,4 +1259,99 @@ public Token getToken() {
     public Logout getLogout() {
         return logout;
     }
+
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java:1262]: It seems to belong the utility code, some properties here can not be even supported by this builder but by various sub-builders

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java [hunk 3]
```diff
@@ -1730,6 +1737,8 @@ public enum ResponseMode {
          */
         public Optional<Boolean> userInfoRequired = Optional.empty();
 
+        private AtomicBoolean dynamicallyEnabledUserInfo = null;
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:1740]: What case does it attempt to support ? We can do it at the build time, but also some other property as far as I recall - like verifying code flow access tokens.

Can we try to avoid having to add some specific code per every such property ?

**Comment by sberyozkin** [comment]
Thanks very much @michalvavrik, appreciate it must've required a lot of patience to deal with so many deprecated properties, I added a couple of comments to start from...
Thanks

**Comment by sberyozkin** [comment]
Hi @michalvavrik IMHO, dealing with the possible modifications of the shared OIDC config from custom resolvers can better be done in a follow up PR as it is a rather edge case, to make it easier to review. 
Replacing deprecated fields with getters is a nice, but also large update on its own

Thanks

**Comment by michalvavrik** [comment]
> Hi @michalvavrik IMHO, dealing with the possible modifications of the shared OIDC config from custom resolvers can better be done in a follow up PR as it is a rather edge case, to make it easier to review. Replacing deprecated fields with getters is a nice, but also large update on its own
> 
> Thanks

What can be done separately is replacing fields access with getters, I'll open PR right now. The rest, which IMHO is vital cannot be done with less code. We can deal with that later.

Whether it is the edge case or not, I don't think we know what users do and we can only be safe if config is immutable, there must be 2 places where OIDC config comes in (one per dynamic, one per static) and one place where the config is stored - that is just my opinion but I think we will avoid troubles. Just the fact that it took me more than 3 days to make this simple PR should itself point out that reusing class instances provided by users can be confusing.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java (hunk 3 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:1740]: @michalvavrik 
>because you don't know whether OIDC Provider supports user info until you have metadata. You cannot do it until you receive response from OIDC provider (if discovery is enabled).

But how is it relevant to what this PR is about ? The user's intention to require it in properties is not related to whether it can actually be satisfied. Furthemore, with OAuth2 providers, no userInfo endpoint exists.

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:1740]: > But how is it relevant to what this PR is about ? The user's intention to require it in properties is not related to whether it can actually be satisfied.

Look here https://github.com/quarkusio/quarkus/blob/b11cdaf9a39f2f1bd4e6ef11a3a5b01d0ad74360/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantContextFactory.java#L550. We only require user info if we detect a CDI injection point and provider supports userInfo.

> Furthemore, with OAuth2 providers, no userInfo endpoint exists.

I don't understand this note. Can you clarify why is it relevant? I suspect you are talking about something else than this code does.

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:1740]: in other words, the line you are commenting on exists so that I can migrate this line https://github.com/quarkusio/quarkus/blob/b11cdaf9a39f2f1bd4e6ef11a3a5b01d0ad74360/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantContextFactory.java#L550.

**Comment by michalvavrik** [comment]
Okay, I think this PR is ready for a review.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java:1262]: I hope I am not confusing you, but let me put it this way: with this PR we control all the entry points and if user changes `OidcTenantConfig` instance after they passed it to Quarkus OIDC code, it has no effect. Which means after we have immutable copy where we control these 2 changeable fields (user info and tenant enabled), we can only rely on methods and state of former config class fields will not be changed (because Quarkus OIDC won't do it and noone else can).

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java (hunk 3 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java]: @michalvavrik This bean contains either discovered metadata or those that is manually configured... In case of OAuth2 providers, ex, GitHub or Strava etc, we point manually to some provider specific account info. Or user may have configured it manually for whatever reasons.

I'm sorry, but I'm really not sure what we are trying to achieve with this UserInfo related updates. IMHO it is not exactly relevant to what this PR plans to do.

**Comment by sberyozkin** [comment]
Hi Michal, I'm probably a little bit confused, but I have an impression there is too much going on in this PR right now, but we can kep discussing. Usually, it is easier for me to review if the PR handles one specific thing at a time

**Comment by sberyozkin** [comment]
Given the PR description:

>avoid using setters and fields that are deprecated and marked for removal

Right, let's do it

>limits modification of "tenant config" to 2 changes (user info and tenant enabled, both are called once at max)

What is the real necessity ? It is only us who can do it in the recorder or in some other code, users can only do it with builders, if they want to call the builder's `tenantId` twice let them do it, etc...

>dynamic tenant config can only by changed from TenantConfigResolver, not anywhere else and users cannot change config state used in a different HTTP requests (which happened in some of our tests)

Like I said, this is such an edge case that one can only do it in the test that I'm not sure the complexity associated with trying to prevent it is necessary. It is just a test hack that I did

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java (hunk 2 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java:1262]: @michalvavrik 

>with this PR we control all the entry points and if user changes OidcTenantConfig instance after they passed it to Quarkus OIDC code, it has no effect.

The fact the test provider did it does not mean anyone would actually want to do it in the real application, why would they. And if they do, may be should fail instead, or log a message that the change happened and attempt to adapt... But it can bring so much complexity, without an obvious benefit, that I'm doubting we should do it

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java]: Please have a look again and tell me how is it not relevant:

1. if https://github.com/quarkusio/quarkus/blob/b11cdaf9a39f2f1bd4e6ef11a3a5b01d0ad74360/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantContextFactory.java#L550 passes
2. https://github.com/quarkusio/quarkus/blob/b11cdaf9a39f2f1bd4e6ef11a3a5b01d0ad74360/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantContextFactory.java#L551 is invoked
3. https://github.com/quarkusio/quarkus/blob/b11cdaf9a39f2f1bd4e6ef11a3a5b01d0ad74360/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantContextFactory.java#L367 user info field can be modified

How can I ignore it? I think there is serious misunderstanding and I am not sure how to explain situation differently.

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java]: That ^ is current main.

**Comment by michalvavrik** [comment]
> > limits modification of "tenant config" to 2 changes (user info and tenant enabled, both are called once at max)
> What is the real necessity ? It is only us who can do it in the recorder or in some other code, users can only do it with builders, if they want to call the builder's `tenantId` twice let them do it, etc...

What this PR does is to limit modifications once the config is final. it is to keep state consistent among whole Quarkus OIDC. What you describe with `tenantId` is something about builder, users can do it million times, this PR doesn't limit anything like this.

I am sorry if I am repeating myself for a millionth time, but what this PR does is makes sure we have 2 places that accepts configuration from either SR Config or from users. And once we accepted that config, we know that this config is final. What is wrong about that? Every other comment you had so far seems completely unrelated to this PR.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java (hunk 3 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java]: @michalvavrik But this is done by the recorder, or as part of the dynamic tenant creation - it is modified like any other field can be. Like I said, you may have a point and we need to tighten it here, let's just have a targeted issue/discussion/PR, you may be right that we are talking about different things, so let's have target each of this issues individually for us to come to the common understanding

**Comment by michalvavrik** [comment]
Got it @sberyozkin . I'll follow your suggestion in smaller dedicated PRs. I'll go back and address your comments independently, but I think there is too much of a discussion in this PR.

I'll close this one and we will start with brand new smaller PR. Thanks @sberyozkin

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java]: > it is modified like any other field can be

just for the record, this is the only scenario that needs to be modified after the config is accepted by Quarkus OIDC, every other modification is done once, when Quarkus OIDC gets `OidcTenantConfig` instance from either user or SR Config.
